### PR TITLE
[Inductor] Add speculative divisibility for Triton kernel dispatch

### PR DESCRIPTION
**Pull request by liqiangxl**
## Summary

Implements the kernel-level runtime dispatch approach discussed in #176653.

### Problem

Under `dynamic=True`, symbolic `SizeArgs` (strides, numels) cannot be statically proven as divisible by 16, so Triton never emits `tt.divisibility=16` annotations and falls back to scalar loads instead of vectorized loads, causing 1.5x–3.4x regressions across pointwise, reduction, and normalization kernels.

#176653 attempted to fix this with Dynamo guards (`expect_true(Eq(expr % 16, 0))`), but as @eellison pointed out, guards are evaluated at the **graph level** — with N dynamic shapes across kernels in a graph, this creates up to 2^N graph recompilations. The correct fix is **kernel-level dispatch**: **compile multiple Triton configs per kernel and dispatch at runtime,** so alignment decisions are local to each kernel and don't pollute the graph-level guard set.

### Solution

AOT-compile two Triton kernel variants per kernel and emit runtime `if/else` dispatch in the wrapper:

- `_div16`: all symbolic SizeArgs speculatively annotated as `tt.divisibility=16`
- `_general`: no speculation (the baseline)

The wrapper emits:
```python
if (a | b | ...) % 16 == 0:
    kernel_div16.run(args, stream=stream0)
else:
    kernel_general.run(args, stream=stream0)
```

where `a, b, ...` are the deduplicated runtime values of speculated SizeArgs. The bitwise OR trick works because 16 is a power of 2: `(a | b) % 16 == 0` is the same as `a % 16 == 0 and b % 16 == 0`.

### Design decisions
- **Gated behind `triton.speculative_divisibility`** (default off, `TORCHINDUCTOR_SPECULATIVE_DIVISIBILITY=1`)
- **Disabled for cpp_wrapper (AOTI)**: uses offline compilation with known shapes
- **Disabled for `autotune_at_compile_time`**: variant dispatch doesn't handle autotune example args generation
- **Disabled for `WrapperFxCodegen`**: FX IR cannot represent if/else dispatch
- **Source reuse**: variant sources generated by repr-replacing `triton_meta` in the fast variant's source, avoiding full codegen per variant
- **Exactly 2 variants** per kernel (fast + fallback) as discussed in #176653 — avoids combinatorial explosion while capturing the common case

### Known limitation / follow-up

The current implementation speculatively annotates **all** symbolic SizeArgs, but not all of them actually influence vectorization. For example, in an inner reduction kernel, `xnumel` (outer grid bound) is irrelevant to vectorization — only `ks0` (stride) and `r0_numel` (reduction bound) matter. Including irrelevant args in the OR condition causes false negatives: `T[17, 1024].sum(dim=1)` falls back to scalar loads because `17 % 16 != 0`, even though all vectorization-relevant args are div16.

A follow-up should filter speculative args to only those that affect the vectorized dimension's mask or address, per kernel type:

## Test
- [x] New `TestSpeculativeDivisibility` class with 6 test cases:
  - Pointwise 1D (2D inputs collapsed single xnumel)
  - Pointwise 2D with mismatched layouts (transposed)
  - 3D middle reduction (T[iter, redu, iter])
  - Inner reduction (T[iter, redu])
  - Outer reduction (T[redu, iter])
  - RMSNorm fused kernel
- [x] Each test verifies dispatch condition shape (pipe count in the `|` expression) and runtime path selection (div16 vs general).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/codegen/triton_utils.py [hunk 1]
```diff
@@ -309,3 +309,69 @@ def is_aligned(x: KernelArgType, alignment: int, include_tensor: bool) -> bool:
...
+    if hasattr(attrs_config, "divisible_by_16"):
+        return set(attrs_config.divisible_by_16)
+    if hasattr(attrs_config, "divisibility_16"):
+        return set(attrs_config.divisibility_16)
+    if isinstance(attrs_config, dict):
+        return {k[0] for k in attrs_config if isinstance(k, tuple)}
+    return set()
+
+
+def config_with_speculative_divisible(
```
Comment (eellison) [inline torch/_inductor/codegen/triton_utils.py:330]: So, if you benchmark:

```
import torch
import triton

@torch.compile()
def foo(x):
    return x * 3.

inp = torch.rand([8192, 73728], device="cuda", dtype=torch.float16)

print(triton.testing.do_bench(lambda: foo(inp)))



@torch.compile()
def foo(x):
    return x * 3.

torch._dynamo.reset()
torch._dynamo.mark_dynamic(inp, 0)
print(triton.testing.do_bench(lambda: foo(inp)))
```

We get the same perf. We don't necessarily need all the symbols to be dynamic, we just to be able to prove that overall numel is divisible by 16. Can we do something similar ? Dispatch based on the overall divisibility required to prove a load will be divisible?

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -5838,7 +5895,20 @@ def add_constexpr_arg(arg_name):
         if config.benchmark_kernel:
             code.splice(self.codegen_kernel_benchmark(num_gb))
 
-        return code.getvalue()
+        src_code = code.getvalue()
+
+        # Generate per-variant Triton source by replacing triton_meta in the source.
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:5898]: One question here is whether or not autotuning for one kernel should set the configs for the other. I sort of doubt different block sizes / num warps will be chosen, although its possible.

**Review by eellison — COMMENTED** [review]
Cool! One comment on more selectively dispatching. Also cc @shunting314 , @v0i0 , @jansel

File: torch/_inductor/codegen/triton_utils.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/codegen/triton_utils.py:330]: The other thing we could consider is if/else around the load and store.. 

```
@triton.jit
def triton_poi_fused_mul_0(in_ptr0, out_ptr0, ks0, ks1, xnumel, XBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x0 = xindex
    if xoffset + XBLOCK <= xnumel:
        tmp0 = tl.load(in_ptr0 + (x0)).to(tl.float32)
    else:
        tmp0 = tl.load(in_ptr0 + (x0), xmask).to(tl.float32)
    tmp1 = tl.full([1], 3.0, tl.float32)
    tmp2 = tmp0 * tmp1
    tmp3 = ks0
    tmp4 = tmp3.to(tl.float32)
    tmp5 = tmp2 * tmp4
    tmp6 = ks1
    tmp7 = tmp6.to(tl.float32)
    tmp8 = tmp5 * tmp7
    if xoffset + XBLOCK <= xnumel:
        tl.store(out_ptr0 + (x0), tmp8)
    else:
        tl.store(out_ptr0 + (x0), tmp8, xmask)
''', device_str='cuda')

Comment (eellison) [reply torch/_inductor/codegen/triton_utils.py:330]: this also gets the same perf.

Comment (liqiangxl) [reply torch/_inductor/codegen/triton_utils.py:330]: This seems to be a better approach and also support partial vectorization. Let me check how to implement it.

Comment (liqiangxl) [reply torch/_inductor/codegen/triton_utils.py:330]: Hi @eellison, it works for pointwise, but for reduction it doesn't improve performance since the stride `ks0` is involved in the load op, e.g. `tmp0 = tl.load(in_ptr0 + (r0_1 + ks0*x0), r0_mask & xmask, eviction_policy='evict_last', other=0.0).to(tl.float32)` without knowing `ks0 % 16 == 0` there doesn't seem exist a safe way to issue vectorized load from triton.

Comment (eellison) [reply torch/_inductor/codegen/triton_utils.py:330]: this is for all reductions, or any particular one ? I think it would still be worth landing the less invasive change for pointwise since I do see (dynamic=True) on pointwise patterns a non-trivial amount.

What do you think?

Comment (liqiangxl) [reply torch/_inductor/codegen/triton_utils.py:330]: It's from a inner reduction `x / x.sum(1, keepdim=True), where x = [M,N]`, similar pattern also exists for outer reduction `x / x.sum(0, keepdim=True)`.

Sure, we can solve the pointwise kernel using this approach. I'll do it in a separate PR.

Comment (liqiangxl) [reply torch/_inductor/codegen/triton_utils.py:330]: Hi @eellison I added https://github.com/pytorch/pytorch/pull/178318
